### PR TITLE
Fixed integrate(1/(2**(2*x/3)+1), (x,0,oo))

### LINKED ISSUE #15810: integrate(1/(2**(2*t/3)+1), (t,0,oo)) is wrong
`integrate(1/(2**(2*x/3)+1), (x,0,oo))` returns 1/2 while [it should be 3/2](https://www.wolframalpha.com/input/?i=Integrate%5B1%2F(2**(2*x%2F3)%2B1),+0,+inf%5D). The indefinite integral is also wrong.

Tested against 1.3 and master.

### PR DESCRIPTION
**Pull request by jmig5776**
#### References to other Issues or PRs

Fixes #15810 

#### Brief description of what is fixed or changed
Earlier the following integral was evaluating wrong.
```
integrate(1/(2**(2*x/3)+1), (x,0,oo))
1/2
```
Now it gives the correct answer i.e 3/2
```
integrate(1/(2**(2*x/3)+1), (x,0,oo))
3/2
```

#### Release Notes

* integrals
  * fixed a bug in `hermite_reduce` function in `risch`.

### CODE REVIEW

**Comment by jksuom** [comment]
It seems that `hermite_reduce` was planned to work with polynomials over a field. A constant `gcd` would then be 1.

**Comment by jmig5776** [comment]
Let me explain in detail
Earlier the ```doit ```function was evaluating the antideriv wrong i.e ```x/3 - log(2**(2*x/3) + 1)/(2*log(2))```
whereas the antideriv should be ```x - 3*log(2**(2*x/3) + 1)/(2*log(2))``` .

Then on further tracking down I got that the ```result``` was coming from ```result, i = risch_integrate(f, x, separate_integral=True, conds=conds)``` Then in ```risch_integrate``` I found out that in ```exp``` case the function integrate_hyperexponential the ans was returning to be ```-log(exp(2*x*log(2)/3) + 1)/(2*log(2))``` which was wrong as this ```-3*log(exp(2*x*log(2)/3) + 1)/(2*log(2))``` was expected.

Then in ```integrate_hyperexponential``` function, hermite_reduce was returning wrong value of h i.e ```(Poly(1, _t0, domain='QQ'), Poly(3*_t0 + 3, _t0, domain='QQ'))``` and the correct value that hermite_reduce should return is ```(Poly(1, _t0, domain='QQ'), Poly(_t0 + 1, _t0, domain='QQ'))```.

Then I found out that in ```hermite_reduce```  while evaluating gcd of ```Poly(_t0 + 1, _t0, domain='ZZ')``` and ```Poly(2*log(2)/3*_t0, _t0, domain='QQ[log(2)]')``` it was returning ```1/3``` while on reading about risch_integrate I got that the moto of gcd is to find gcd of polynomial with denominator , So I added that if  gcd has degree zero in variable ```DE.t``` then it should return ```S.One```. And then it solved the problem.

**Comment by jmig5776** [comment]
@asmeurer , @jksuom I had added the test. Is the fix is Okk?

**Comment by jmig5776** [comment]
Sorry, I accidently clicked on the merge, while resolving the conflict. What should I do now ?

**Comment by jmig5776** [comment]
I had added test case successfully, please review it .

**Comment by asmeurer** [comment]
Can you add a direct test for hermite_reduce as well?

**Comment by asmeurer** [comment]
I agree that virtually every gcd in the risch algorithm should use "field gcd". Can we fix this by making sure the polynomials in question have a field domain? Where are `Poly(_t0 + 1, _t0, domain='ZZ')` and `Poly(2*log(2)/3*_t0, _t0, domain='QQ[log(2)]')` created?

**Comment by jmig5776** [comment]
> Can you add a direct test for hermite_reduce as well?

Okk i will add it.

**Comment by jksuom** [comment]
The domains of the input polynomials can probably be made fields by starting with

    l = Poly(1/d.LC(), DE.t, field=True)

but that is not enough, the domains have to be passed over to derived polynomials as well. (Maybe something like `...as_poly(t, domain=p.domain)` in `derivation`.)

**Comment by jmig5776** [comment]
@jksuom , I will try it as you suggested and how to find "field gcd"?

**Comment by asmeurer** [comment]
Sorry for the ambiguity. By "field gcd" I mean that `gcd` behaves differently if the domain of the input is a field or not. The behavior when the domain is a field is generally what we want in risch. The Risch algorithm works with fields everywhere, except the top level of the tower where it deals with polynomials  (numerator and denominator separately). 

This is not the first time this sort of bug has occurred in risch. I wonder if there's a more generic way to fix it so we can avoid it happening again.

**Comment by asmeurer** [comment]
You have a minor formatting issue in your release notes entries. You are indenting the second bullet point too far (notice how it shows as a `*` instead of an actual bullet `•`). This is causing the bot to format it wrong with an extra newline at the end before the link back to the PR and author. In markdown you should only add two more spaces to indent a bullet point. 

Also, you do not need to use three backticks unless the code is multiple lines. 

That is, instead of 

```
* integrals
        * fixed a bug in ```hermite_reduce``` function in ```risch```.
```

Just write

```
* integrals
  * fixed a bug in `hermite_reduce` function in `risch`.
```

(I have already fixed this one, but other PRs may need fixing as well)

**Comment by jmig5776** [comment]
> Sorry for the ambiguity. By "field gcd" I mean that `gcd` behaves differently if the domain of the input is a field or not. The behavior when the domain is a field is generally what we want in risch. The Risch algorithm works with fields everywhere, except the top level of the tower where it deals with polynomials (numerator and denominator separately).
> 
> This is not the first time this sort of bug has occurred in risch. I wonder if there's a more generic way to fix it so we can avoid it happening again.

I think @skirpichev had implemented as you described.

**Comment by jmig5776** [comment]
> You have a minor formatting issue in your release notes entries.
> (I have already fixed this one, but other PRs may need fixing as well)

Sorry, it won't happen again.

**Comment by jksuom** [comment]
It appears that this will fix the issue at hand, but I think that we could try to use field coefficients more extensively. That could probably be implemented by making `field=True` the default in `frac_in` by adding

    if 'field' not in kwargs:
        kwargs['field'] = True

It would be interesting to see how many tests would have to be edited. There are going to be changes like `ZZ -> QQ` but hopefully nothing worse than that.

If that will work, then I should rather have `derivation` fixed by adding `domain=p.domain` instead of `field=True` (expecting that `p.domain` would be a field, in general).

**Comment by jmig5776** [comment]
Okk @jksuom

**Comment by jmig5776** [comment]
@jksuom Can you please review it?.

File: sympy/integrals/tests/test_risch.py [hunk 1]
```diff
@@ -707,15 +707,6 @@ def test_DifferentialExtension_equality():
...
-        "('cases', ['base', 'exp', 'primitive']), ('case', 'primitive'), ('t', t1), "
-        "('d', Poly(2*t0*x/(t0 + 1), t1, domain='ZZ(x,t0)')), ('newf', t0**2 + t1), ('level', -1), "
-        "('dummy', False)]))")
-
-    assert str(DE) == ("DifferentialExtension({fa=Poly(t1 + t0**2, t1, domain='ZZ[t0]'), "
-        "fd=Poly(1, t1, domain='ZZ'), D=[Poly(1, x, domain='ZZ'), Poly(2*x*t0, t0, domain='ZZ[x]'), "
-        "Poly(2*t0*x/(t0 + 1), t1, domain='ZZ(x,t0)')]})")
+    assert repr(DE) == ("DifferentialExtension(dict([('f', exp(2*x**2) + log(exp(x**2) + 1)), ('x', x), ('T', [x, t0, t1]), ('D', [Poly(1, x, domain='ZZ'), Poly(2*x*t0, t0, domain='ZZ[x]'), Poly(2*t0*x/(t0 + 1), t1, domain='ZZ(x,t0)')]), ('fa', Poly(t1 + t0**2, t1, domain='ZZ(t0)')), ('fd', Poly(1, t1, domain='QQ')), ('Tfuncs', [Lambda(i, exp(i**2)), Lambda(i, log(t0 + 1))]), ('backsubs', []), ('exts', [None, 'exp', 'log']), ('extargs', [None, x**2, t0 + 1]), ('cases', ['base', 'exp', 'primitive']), ('case', 'primitive'), ('t', t1), ('d', Poly(2*t0*x/(t0 + 1), t1, domain='ZZ(x,t0)')), ('newf', t0**2 + t1), ('level', -1), ('dummy', False)]))")
```
Comment (jksuom) [inline sympy/integrals/tests/test_risch.py]: These long lines should be split.

Comment (jmig5776) [reply sympy/integrals/tests/test_risch.py]: Okk Okk

**Comment by jmig5776** [comment]
If we give `domain=p.domain` as argument then this error is coming `AttributeError: 'Symbol' object has no attribute 'domain'`

**Comment by jksuom** [comment]
The docstring of `derivation` says that "p is a polynomial in t over K(x)", and polynomials should have attribute `domain`. Is it possible that `p` is not a polynomial? Under which conditions does the attribute error appear?

**Comment by jmig5776** [comment]
I also went through `derivation` . But in most of the test cases it is failing in my travis tests. One of them is this
```
Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/sympy-1.4.dev0-py3.6.egg/sympy/utilities/runtests.py", line 1306, in _timeout
    function()
  File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/sympy-1.4.dev0-py3.6.egg/sympy/integrals/tests/test_risch.py", line 660, in test_risch_integrate
    ans1 = risch_integrate(e1, x)
  File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/sympy-1.4.dev0-py3.6.egg/sympy/integrals/risch.py", line 1757, in risch_integrate
    ans, i, b = integrate_primitive(fa, fd, DE)
  File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/sympy-1.4.dev0-py3.6.egg/sympy/integrals/risch.py", line 1404, in integrate_primitive
    q, i, b = integrate_primitive_polynomial(p, DE)
  File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/sympy-1.4.dev0-py3.6.egg/sympy/integrals/risch.py", line 1356, in integrate_primitive_polynomial
    rv = limited_integrate(aa, ad, [(Dta, Dtb)], DE)
  File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/sympy-1.4.dev0-py3.6.egg/sympy/integrals/prde.py", line 809, in limited_integrate
    h, A = param_rischDE(Fa, Fd, G, DE)
  File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/sympy-1.4.dev0-py3.6.egg/sympy/integrals/prde.py", line 709, in param_rischDE
    h, B = param_poly_rischDE(a, b, r, n, DE)
  File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/sympy-1.4.dev0-py3.6.egg/sympy/integrals/prde.py", line 537, in param_poly_rischDE
    return prde_cancel_liouvillian(b, q, n, DE)
  File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/sympy-1.4.dev0-py3.6.egg/sympy/integrals/prde.py", line 459, in prde_cancel_liouvillian
    ba, bd = frac_in(b + i*derivation(DE.t, DE)/DE.t,
  File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/sympy-1.4.dev0-py3.6.egg/sympy/integrals/risch.py", line 899, in derivation
    r += (d*pv.diff(v)).as_poly(t, field=p.domain.is_Field)
AttributeError: 'Symbol' object has no attribute 'domain'
```

**Comment by jksuom** [comment]
File "/home/travis/virtualenv/python3.6.3/lib/python3.6/site-packages/sympy-1.4.dev0-py3.6.egg/sympy/integrals/prde.py", line 459, in prde_cancel_liouvillian
    ba, bd = frac_in(b + i*derivation(DE.t, DE)/DE.t,

Here `derivation(DE.t, DE)/DE.t` looks like a bug. I think it should probably be `DE.d.quo(Poly(DE.t, DE.t)`.

**Comment by jmig5776** [comment]
I tried it but Unification error is occuring.
```
File "/home/jmig5776/Gsoc/sympy/sympy/integrals/tests/test_risch.py", line 660, in test_risch_integrate
    ans1 = risch_integrate(e1, x)
  File "/home/jmig5776/Gsoc/sympy/sympy/integrals/risch.py", line 1757, in risch_integrate
    ans, i, b = integrate_primitive(fa, fd, DE)
  File "/home/jmig5776/Gsoc/sympy/sympy/integrals/risch.py", line 1404, in integrate_primitive
    q, i, b = integrate_primitive_polynomial(p, DE)
  File "/home/jmig5776/Gsoc/sympy/sympy/integrals/risch.py", line 1356, in integrate_primitive_polynomial
    rv = limited_integrate(aa, ad, [(Dta, Dtb)], DE)
  File "/home/jmig5776/Gsoc/sympy/sympy/integrals/prde.py", line 808, in limited_integrate
    h, A = param_rischDE(Fa, Fd, G, DE)
  File "/home/jmig5776/Gsoc/sympy/sympy/integrals/prde.py", line 708, in param_rischDE
    h, B = param_poly_rischDE(a, b, r, n, DE)
  File "/home/jmig5776/Gsoc/sympy/sympy/integrals/prde.py", line 536, in param_poly_rischDE
    return prde_cancel_liouvillian(b, q, n, DE)
  File "/home/jmig5776/Gsoc/sympy/sympy/integrals/prde.py", line 459, in prde_cancel_liouvillian
    ba, bd = frac_in(b + i*DE.d.quo(Poly(DE.t, DE.t)))
  File "/home/jmig5776/Gsoc/sympy/sympy/core/decorators.py", line 91, in __sympifyit_wrapper
    return func(a, b)
  File "/home/jmig5776/Gsoc/sympy/sympy/polys/polytools.py", line 4010, in __add__
    return f.add(g)
  File "/home/jmig5776/Gsoc/sympy/sympy/polys/polytools.py", line 1341, in add
    _, per, F, G = f._unify(g)
  File "/home/jmig5776/Gsoc/sympy/sympy/polys/polytools.py", line 404, in _unify
    dom, lev = f.rep.dom.unify(g.rep.dom, gens), len(gens) - 1
  File "/home/jmig5776/Gsoc/sympy/sympy/polys/domains/domain.py", line 262, in unify
    return K0.unify_with_symbols(K1, symbols)
  File "/home/jmig5776/Gsoc/sympy/sympy/polys/domains/domain.py", line 240, in unify_with_symbols
    raise UnificationFailed("can't unify %s with %s, given %s generators" % (K0, K1, tuple(symbols)))
sympy.polys.polyerrors.UnificationFailed: can't unify QQ(x) with ZZ, given (_t0, x) generators
```

**Comment by asmeurer** [comment]
What are the polynomials in question? `x` should not be a generator if it is working at the level of `_t0`. Risch should in general work only with polynomials of one generator.

**Comment by jksuom** [comment]
It is computing this sum `b + i*DE.d.quo(Poly(DE.t, DE.t)`. What is `b`? Does it contain `x`?

**Comment by jmig5776** [comment]
b is `Poly(0, _t0, domain='QQ(x)')`

**Comment by asmeurer** [comment]
This is in a DecrementLevel block, so I guess _t0 is what should not be there.

**Comment by jmig5776** [comment]
So what should be done to remove it.

**Comment by jksuom** [comment]
I suspect that DecrementLevel is wrong in this context. That has to be investigated more closely. For now, I suggest that the [call](https://github.com/sympy/sympy/blob/master/sympy/integrals/prde.py#L537) be temporarily replaced by a NotImplementedError.

**Comment by jmig5776** [comment]
Okk I had added Not Implemented error for now.

**Comment by asmeurer** [comment]
Also I think the page numbers in the comments of that function are wrong. They reference an equation at the top of page 238 but page 238 has no equations on it. Perhaps they were referring to the PDF page number instead of the real page number from the book? @gxyd

**Comment by jksuom** [comment]
The algorithm is described on p. 237 in the book: The Liouvillean Case, but there is no pseudocode. That makes it harder to see which should be the active level of DE.

**Comment by jksuom** [comment]
The book is
 Manuel Bronstein, Symbolic Integration I, Transcendental Functions, Springer

It explains an implementation of Risch's algorithm for purely transcendental functions. Most of that is coded in risch.py, rde.py and prde.py, almost all of those routines for which there is pseudocode given in the book. Some parts of the implementation, such as the Liouvillian case at hand, are only verbally described and currently not completely implemented and tested. However, I think that there is no need to include them in this PR. That is why I suggested temporarily raising NotImplementedError. The work can be continued later.

File: sympy/integrals/risch.py [hunk 2]
```diff
@@ -895,7 +896,7 @@ def derivation(p, DE, coefficientD=False, basic=False):
         if basic:
             r += d.as_expr()*pv.diff(v)
         else:
-            r += (d*pv.diff(v)).as_poly(t)
+            r += (d*pv.diff(v)).as_poly(t, field=p.domain.is_Field)
```
Comment (jksuom) [inline sympy/integrals/risch.py]: This should probably be `field=True`.

Comment (jmig5776) [reply sympy/integrals/risch.py]: Okk @jksuom . I had commited it in my new commit.

Sorry for the late response, my exams were going previous week.

**Comment by oscarbenjamin** [comment]
What's the current status of this PR?

**Comment by czgdp1807** [comment]
@jmig5776 Are you still working on it? If so, then please resolve the conflicts. Please update.

**Comment by jmig5776** [comment]
@czgdp1807 No I am closing the PR.

**Pull request closed without merging**